### PR TITLE
feat: Support Broadcast HashJoin

### LINKED ISSUE #202: Support BroadcastHashJoinExec
### What is the problem the feature request solves?

In our benchmark of TPCDS queries, `BroadcastHashJoinExec` is shown frequently in the query plans. To improve Comet coverage, we should support `BroadcastHashJoinExec`.

This is blocked by #193.

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #202.

### CODE REVIEW

**Review by viirya — COMMENTED** [review]
This is based on #194.

File: common/src/main/scala/org/apache/spark/sql/comet/execution/shuffle/ArrowReaderIterator.scala
Comment (viirya) [inline common/src/main/scala/org/apache/spark/sql/comet/execution/shuffle/ArrowReaderIterator.scala:45]: We need to release the batch before loading next batch. Because `ArrowStreamReader` loads data into same vectors of root internally. After loading next batch, `close` will release the just loaded batch instead of previous batch.

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBroadcastExchangeExec.scala
Comment (viirya) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometBroadcastExchangeExec.scala:262]: We iterates each batch and close previous one. `flatMap` will iterates all batches then creates new iterator on them. So it will corrupt these batches.

Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBroadcastExchangeExec.scala:262]: We need to construct a custom iterator and manually iterates these batches one by one.

**Comment by viirya** [comment]
cc @sunchao

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBroadcastExchangeExec.scala
Comment (sunchao) [inline spark/src/main/scala/org/apache/spark/sql/comet/CometBroadcastExchangeExec.scala:272]: is it possible that this `batchesIter.next` has no more next but the next item in `batchesIter` still has more elements? e.g.: `[1, 2, 3], [], [4, 5, 6]`?

File: common/src/main/scala/org/apache/spark/sql/comet/execution/shuffle/ArrowReaderIterator.scala
Comment (sunchao) [reply common/src/main/scala/org/apache/spark/sql/comet/execution/shuffle/ArrowReaderIterator.scala:45]: Is this related to the memory leak we saw?

Comment (viirya) [reply common/src/main/scala/org/apache/spark/sql/comet/execution/shuffle/ArrowReaderIterator.scala:45]: > Is this related to the memory leak we saw?

It's not, although I suspected it before too. For shuffle, a channel only contains one batch, so `ArrowReaderIterator` doesn't hit this issue.

Comment (viirya) [reply common/src/main/scala/org/apache/spark/sql/comet/execution/shuffle/ArrowReaderIterator.scala:45]: > This sounds like a data corruption problem. If the just loaded batch is closed/released, the just loaded ColumnarBatch would be corrupted? But it seems like that the CI passes without any issue previously.
> 
> When working on #206, I also found out it might be inconvenient to use Arrow Java's memory API. It requires extra caution to allocate and release ArrowBuf correctly.

Due to https://github.com/apache/arrow-datafusion-comet/pull/211#discussion_r1535661988, this issue is not exposed before.

I feel that Arrow Java API is hard to use and somehow counter-intuitive, especially compared with arrow-rs.

Comment (sunchao) [reply common/src/main/scala/org/apache/spark/sql/comet/execution/shuffle/ArrowReaderIterator.scala:45]: Yes I feel the same pain when using Java Arrow. I think in the long term we'd better to switch away from it. It should be relatively easy except the Java Arrow Flight feature.

File: common/src/main/scala/org/apache/comet/vector/NativeUtil.scala
Comment (viirya) [reply common/src/main/scala/org/apache/comet/vector/NativeUtil.scala:57]: If the batch has its provider, it should be returned in `batchProviderOpt `?

File: spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala
Comment (viirya) [reply spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala:338]: Yea

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBroadcastExchangeExec.scala
Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBroadcastExchangeExec.scala:262]: Yea

File: spark/src/main/scala/org/apache/spark/sql/comet/operators.scala
Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/operators.scala:688]: yes

File: spark/src/main/scala/org/apache/spark/sql/comet/CometBroadcastExchangeExec.scala
Comment (viirya) [reply spark/src/main/scala/org/apache/spark/sql/comet/CometBroadcastExchangeExec.scala]: I restored this part.

File: common/src/main/scala/org/apache/comet/vector/NativeUtil.scala
Comment (viirya) [reply common/src/main/scala/org/apache/comet/vector/NativeUtil.scala:57]: Oh I see. I suppose that the dictionary provider is same across batches. This seems to be the reason why there is dictionary provider, i.e. to store dictionary values for arrays/batches.

**Comment by viirya** [comment]
@sunchao any more comments?

**Review by sunchao — APPROVED** [review]
LGTM. Can we merge #210 first and trigger another CI run?

**Comment by viirya** [comment]
Let me take a look at #210 tomorrow (today is late, I might not be able to finish it).

I'll hold this until #210 is merged.

**Comment by viirya** [comment]
Merged. Thanks.

**Pull request merged**
